aws/product/awsec2product.go

```diff
@@ -0,0 +1,191 @@
+	if value, ok := value.(string); ok {
+		switch name {
+		case "instanceType":
+			dbAwsProduct.InstanceType = value
+		case "memory":
+			dbAwsProduct.Memory = value
+		case "storage":
+			dbAwsProduct.Storage = value
+		case "networkPerformance":
+			dbAwsProduct.NetworkPerformance = value
+		case "tenancy":
+			dbAwsProduct.Tenancy = value
+		case "operatingSystem":
+			dbAwsProduct.OperatingSystem = value
+		}
+	}
+}
+
+// importResult parses the body returned by downloadJSON and
+// inserts the pricing in the database.
+func importResult(reader io.ReadCloser, logger jsonlog.Logger) {
+	defer reader.Close()
+	body, err := ioutil.ReadAll(reader)
+	if err != nil {
+		logger.Error("Body reading failed", err)
+	}
+	var bodyMap map[string]interface{}
+	if err = json.Unmarshal(body, &bodyMap); err != nil {
+		logger.Error("JSON Unmarshal failed", err)
+		return
+	}
+	for _, product := range bodyMap["products"].(map[string]interface{}) {
+		var dbAwsProduct models.AwsProductEc2
+		product := product.(map[string]interface{})
+		dbAwsProduct.Sku = product["sku"].(string)
+		attributes := product["attributes"].(map[string]interface{})
+		for name, attribute := range attributes {
+			if fct, ok := storeAttribute[name]; ok {
+				fct(name, attribute, &dbAwsProduct, logger)
+			}
+		}
+		dbAwsProduct.Insert(db.Db)
+	}
+}
+
+// purgeCurrent purges the current pricing from the database.
+func purgeCurrent() {
+	// sql query
+	const sqlstr = `DELETE FROM trackit.aws_product_ec2`
+
+	// run query
+	models.XOLog(sqlstr)
+	db.Db.Exec(sqlstr)
+}
+
+// downloadJSON requests AWS' API and returns the body after
+// checking if there's already the same version in the database.
+func downloadJSON(logger jsonlog.Logger) (io.ReadCloser, error) {
+	hc := http.Client{}
+	req, err := http.NewRequest("GET", urlEC2Pricing, nil)
+	if err != nil {
+		return nil, err
+	}
+	lastFetch, err := models.AwsFetchPricingByProduct(db.Db, "ec2")
+	if err != nil {
+		req.Header.Add("If-None-Match", lastFetch.Etag)
+	}
+	res, err := hc.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	if res.StatusCode == 304 {
+		logger.Info("JSON Downloading halted: Already exists with ETag.\n", lastFetch.Etag)
+		return nil, nil
+	}
+	if lastFetch != nil {
+		lastFetch.Delete(db.Db)
+	}
+	dbAfp := models.AwsFetchPricing{
+		Product: "ec2",
+		Etag:    res.Header["Etag"][0],
+	}
+	err = dbAfp.Save(db.Db)
+	if err != nil {
+		logger.Error("Error during dbAfp saving", err)
+		return nil, err
+	}
+	return res.Body, nil
+}
+
+// ImportEC2Pricing downloads the EC2 Pricing from AWS and
+// store it in the database.
+func ImportEC2Pricing() (interface{}, error) {
+	logger := jsonlog.DefaultLogger
+	res, err := downloadJSON(logger)
+	if err != nil {
+		logger.Error("Error during JSON downloading", err)
+	} else if res != nil {
+		purgeCurrent()
+		importResult(res, logger)
+	}
+	return nil, nil
```

Comment:
Why do you need the `interface{}` return type, if you will always return `nil` ?

Reply:
In that case, why is there even a return value ?